File: backend/app/schemas/message.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.message import MessageSender


class MessageBase(BaseModel):
    conversation_id: int
    sender: MessageSender
    raw_content: str = Field(min_length=1)
    narration: str | None = None
    dialogue: str | None = None
    emotion: str | None = None
    message_metadata: dict[str, Any] = Field(default_factory=dict)
    sequence_number: int = Field(ge=0)


class MessageCreate(MessageBase):
    pass


class MessageUpdate(BaseModel):
    raw_content: str | None = Field(default=None, min_length=1)
    narration: str | None = None
    dialogue: str | None = None
    emotion: str | None = None


class MessageRead(MessageBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    is_edited: bool
    created_at: datetime
    edited_at: datetime | None

File: frontend/src/types/message.ts
export interface MemoryProcessingResult {
  created: number
  consolidated: number
  superseded: number
  rejected: number
  already_applied: number
  memory_ids: number[]
  warnings: string[]
}
export interface Message {
  id: number | string
  conversation_id: number
  sender: 'user' | 'character' | 'system'
  raw_content: string
  narration: string | null
  dialogue: string | null
  emotion: string | null
  message_metadata: Record<string, unknown>
  sequence_number: number
  is_edited: boolean
  created_at: string
  edited_at: string | null
  temporary?: boolean
}
export interface MessageList {
  items: Message[]
  total: number
  limit: number
  offset: number
  has_more: boolean
}
export interface SendMessageRequest {
  content: string
  client_message_id: string
  response_length?: 'concise' | 'balanced' | 'detailed'
  language_mode?: string
}
export interface SendMessageResponse {
  user_message: Message
  character_message: Message
  memory: MemoryProcessingResult | null
  warnings: string[]
}
export interface RegenerateResponse {
  character_message: Message
  turn_count: number
  memory: MemoryProcessingResult | null
  warnings: string[]
}
